Added #[projections(primitive)] attribute for structures.

Alizter opened the pull request:
We move the flag `primitive_flag ()` to `vernacentries.ml` and instead pass a bool for enabling primitive projections. An attribute `#[projections(primitive{={yes|no}})]` is added allowing structures to have primitive projections enabled or disabled locally. By default, `primitive_flag ()` gets passed to the bool values, which appear in the decleration of structures.

Fixes / closes #13150

I've opted for `#[projections(...)]` as `#[primitive]` could be confused with the primitive data structures.

I'm waiting for more feedback before writing the documentation.

- [x] Added / updated **test-suite**.
- [x] Added **changelog**.
- [x] Added / updated **documentation**.
  - [x] Documented any new / changed **user messages**.
  - [x] Updated **documented syntax** by running `make -f Makefile.make doc_gram_rsts`.

Alizter commented on vernac/record.ml:
```diff
@@ -596,7 +590,7 @@ let declare_structure ~cumulative finite ~ubind ~univs ~variances paramimpls par
   let impls = List.map (fun _ -> paramimpls, []) record_data in
   let kn = DeclareInd.declare_mutual_inductive_with_eliminations mie ubind impls
-      ~primitive_expected:(primitive_flag ())
+      ~primitive_expected:primitive_proj
```
Should I rename this label to `primitive_proj` to be more homogeneous?

gares replied:
I'd say so, but maybe in another PR

Alizter commented on vernac/vernacentries.ml:
```diff
@@ -849,17 +849,33 @@ let private_ind =
   | None -> return false
 
+(** Flag governing use of primitive projections. Disabled by default. *)
+let primitive_flag =
```
Not sure where to put this

Alizter commented on vernac/vernacentries.ml:
```diff
@@ -849,17 +849,33 @@ let private_ind =
-  let ((template, (poly, cumulative)), private_ind), typing_flags = Attributes.(
+  let (((template, (poly, cumulative)), private_ind), typing_flags), primitive_proj =
+    Attributes.(
       parse Notations.(
           template
           ++ polymorphic_cumulative ~is_defclass:(Option.has_some is_defclass)
-          ++ private_ind ++ typing_flags)
+          ++ private_ind ++ typing_flags ++ primitive_proj)
```
Is there a better way to do this?

Alizter replied:
It seemed a bit fragile to me but I might be wrong.

Alizter commented on vernac/record.ml:
```diff
@@ -883,15 +877,15 @@ let regular_structure ~cumulative ~template ~ubind ~impargs ~univs ~variances ~p
     list telling if the corresponding fields must me declared as coercions
     or subinstances. *)
-let definition_structure udecl kind ~template ~cumulative ~poly
+let definition_structure udecl kind ~template ~cumulative ~poly ~primitive_proj
```
Unfortunately, we have to pass the `primitive_proj` bool all the way here, even though it's only relevent for classes and other structures.

Alizter commented:
This needs some more work since we can do:
```coq
#[projections(primitive)]
Inductive nat := 
| zero : nat
| succ : nat -> nat
.
```

Which is nonsense.

Alizter commented:
I've added a check in `vernac_inductive` that checks if the `#[projection(primitive)]` attribute is allowed.

gares commented on vernac/vernacentries.ml:
```diff
@@ -849,17 +849,52 @@ let private_ind =
+  in
+  (* We only allow the #[projections(primitive)] attribute
+     in special cases. *)
+  let pp_attr : bool Attributes.Notations.t =
```
```suggestion
  let prim_proj_attr : bool Attributes.Notations.t =
```
pp is used everywhere for pretty printer

gares commented on vernac/vernacentries.ml:
```diff
@@ -849,17 +849,52 @@ let private_ind =
+    else if List.for_all is_constructor indl then
+      Notations.return false
+    else
+      primitive_proj
```
I wonder why this is not `List.for_all is_record` (but I did not try).

gares commented:
To me, this is very a welcome syntax update. CC @coq/doc-maintainers for any feedback before @Alizter edits the doc.

gares commented on vernac/vernacentries.ml:
```diff
@@ -849,17 +849,48 @@ let private_ind =
+  | _ -> false
+  in
+  (* We only allow the #[projections(primitive)] attribute
+     in special cases. *)
```
```suggestion
     for records. *)
```

gares commented on vernac/vernacentries.ml:
```diff
@@ -849,17 +849,48 @@ let private_ind =
+     in special cases. *)
+  let prim_proj_attr : bool Attributes.Notations.t =
+    if List.for_all is_record indl then primitive_proj
+    else Notations.return false
```
I'm not an expert of these parsing combinators, maybe @SkySkimmer can tell us if this is the right way to fail

SkySkimmer replied:
This works OK, it will say something like "this command does not support this attribute: projections".
If you want a more customized error you can do like polymorphic_cumulative, passing a bool to primitive_proj so that it errors in the `Some` case.

Zimmi48 commented:
This sounds like a very good addition indeed. The attribute should probably be documented next to the corresponding flag, and be mentioned in the list of supported attributes of the `Record` command. BTW, I do not feel like the doc of primitive projections is very well organized right now (with the flag description coming before the main presentation of primitive records), so feel free to change that if you want (and feel free not to do it too).

Zimmi48 commented:
Can you rebase your PR to get rid of the merge commit?

Alizter commented:
Sorry about that, I messed up my local master

jfehrle commented on doc/changelog/07-vernac-commands-and-options/14699-alizter+primitive-attribute.rst:
```diff
@@ -0,0 +1,6 @@
+- **Added:**
+  :attr:`projections(primitive)` attribute for marking a record use
+  primitive projections.
```
```suggestion
  :attr:`projections(primitive)` attribute for marking a record using
  primitive projections
```

Trivial picky detail: The period for the last sentence in these (you have only one) is after the `by Ali Caglayan).`, do you don't need one after `projections`

Alizter replied:
Ah OK, this was confusing me since I saw some full stops in the change log elsewhere. In the make-changelog script perhaps we should write
```
Bla Bla Bla (don't end with a period)
```
or something.

Zimmi48 replied:
Yes, please, feel free to open a PR improving this. I've fixed many of these full stops before parentheses but at some point I've got tired and stopped caring about it.

Zimmi48 replied:
This comment was marked as resolved but I don't see the change to the changelog entry.

BTW, I would have fixed "marking a record use primitive projections" differently from Jim's suggestion: I would have suggested:
```suggestion
  :attr:`projections(primitive)` attribute to make a record use
  primitive projections
```

Alizter replied:
Done.

Alizter commented:
I will try to update the documentation this week. Currently the CI is complaining because the changelog is referencing nonexistent attributes.

Alizter commented:
I've added some documentation and done some small refactoring in the `Primitive Projections` section. Notably I removed the term "primitive records" since I don't believe this is a very useful term to mention. We use it nearly nowhere else and it can be confused with the other "primitive datatypes" which have little to do with this. I've also reordered the appearance of the options so that it is a bit clearer.

Alizter commented:
Also would it be possible to autogenerate the list of attributes for these keywords? At the moment we manually have to mention them, but I could very easily have forgotten to mention this new attribute in the classes and coinductive documentation.

Zimmi48 commented:
> Also would it be possible to autogenerate the list of attributes for these keywords? At the moment we manually have to mention them, but I could very easily have forgotten to mention this new attribute in the classes and coinductive documentation.

It would be great if we could. One difficulty will be that there is no central place where attribute lists are collected and stored. But feel free to open an issue to discuss how we could do this.

A related question is auto-generating the list of warnings / error messages. See #7373 and #11901.

Zimmi48 commented on doc/sphinx/language/core/records.rst:
```diff
@@ -235,34 +235,12 @@ the errors of inductive definitions, as described in Section
+When the :flag:`Primitive Projections` flag is on or the 
+:attr:`projections(primitive)` is supplied, definitions of record types change 
+meaning. When a type is declared with primitive projections, its :g:`match` 
+construct is disabled (see :ref:`primitive_projections` though).
```
The comment in the parentheses comes from the previous text, but now it becomes self-referential and useless.

Alizter replied:
Alright, removed.

Zimmi48 commented on doc/sphinx/language/core/records.rst:
```diff
@@ -235,34 +235,12 @@ the errors of inductive definitions, as described in Section
-.. flag:: Primitive Projections
-
-   This :term:`flag` turns on the use of primitive
-   projections when defining subsequent records (even through the ``Inductive``
```
Maybe keeping the mention that this applies even with the records defined with the `Inductive` or `CoInductive` commands would be useful (actually, it is kept in the description of the `flag`...). By the way, you forgot to list the attribute in the supported attributes for the `Inductive` command.

Alizter replied:
I didn't mention it for Inductive since it is technically not supported. I believe only for the `:= { .. }` syntax.

Alizter replied:
Alright, I've modified all the other mentions of the attribute to specify that it is valid only when using record syntax. This includes `Inductive`.

Zimmi48 commented on doc/sphinx/language/core/records.rst:
```diff
@@ -235,34 +235,12 @@ the errors of inductive definitions, as described in Section
+:attr:`projections(primitive)` is supplied, definitions of record types change 
+meaning. When a type is declared with primitive projections, its :g:`match` 
+construct is disabled (see :ref:`primitive_projections` though).
+To eliminate the (co-)inductive type, one must use its defined primitive 
```
Wouldn't it read better like this?
```suggestion
To eliminate the record type, one must use its defined primitive 
```

Zimmi48 commented on doc/sphinx/language/core/records.rst:
```diff
@@ -235,34 +235,12 @@ the errors of inductive definitions, as described in Section
+To eliminate the (co-)inductive type, one must use its defined primitive 
+projections.
 
 .. The following paragraph is quite redundant with what is above
```
Do you still consider this comment to be relevant? It seems to me that it could be removed after your changes.

Alizter replied:
I will keep it as I think it's useful when skimming the documentation.

Zimmi48 commented on doc/sphinx/language/core/records.rst:
```diff
@@ -286,6 +264,37 @@ There are currently two ways to introduce primitive records types:
    elimination is not available for them and only non-dependent case analysis
    can be defined.
 
+For both cases the :term:`flag` :flag:`Primitive Projections` must be set or 
+the :term:`attribute` :attr:`projections(primitive)` must be supplied.
```
```suggestion
For both cases the :flag:`Primitive Projections` :term:`flag` must be set or 
the :attr:`projections(primitive)` :term:`attribute` must be supplied.
```

Zimmi48 commented:
It looks like you have some whitespace errors. You might need to do a `git rebase --whitespace=fix` to solve this.

Zimmi48 commented on doc/sphinx/addendum/type-classes.rst:
```diff
@@ -325,8 +325,10 @@ Summary of the commands
+   :attr:`universes(cumulative)`, :attr:`private(matching)` attributes.
+
+   When record syntax is used, this command also supports the
+   :attr:`projections(primitive)` :term:attribute.
```
```suggestion
   :attr:`projections(primitive)` :term:`attribute`.
```

Zimmi48 commented on doc/sphinx/language/core/coinductive.rst:
```diff
@@ -30,6 +30,9 @@ More information on co-inductive definitions can be found in
    :attr:`bypass_check(positivity)`, and :attr:`using` attributes.
 
+   When record syntax is used, this command also supports the
+   :attr:`projections(primitive)` :term:attribute.
```
```suggestion
   :attr:`projections(primitive)` :term:`attribute`.
```

Zimmi48 commented on doc/sphinx/language/core/inductive.rst:
```diff
@@ -36,6 +36,9 @@ Inductive types
    :attr:`private(matching)` attributes.
 
+   When record syntax is used, this command also supports the
+   :attr:`projections(primitive)` :term:attribute.
```
```suggestion
   :attr:`projections(primitive)` :term:`attribute`.
```

Zimmi48 commented on doc/sphinx/language/core/records.rst:
```diff
@@ -281,22 +256,52 @@ There are currently two ways to introduce primitive records types:
+.. flag:: Primitive Projections
+
+   This :term:`flag` turns on the use of primitive projections when defining
+   subsequent records (even through the ``Inductive`` and ``CoInductive``
```
```suggestion
   subsequent records (even through the :cmd:`Inductive` and :cmd:`CoInductive`
```

Alizter replied:
There were quite a few of these littered around the document, I've changed as many as I could find.

Zimmi48 commented:
@Alizter Did you notice the comment on the changelog?

Alizter commented:
@Zimmi48 I thought I fixed that, let me have another look.

Alizter commented:
@Zimmi48 Alright, fixed the changelog.

jfehrle commented on doc/sphinx/language/core/records.rst:
```diff
@@ -235,68 +236,73 @@ the errors of inductive definitions, as described in Section
-the folded version delta-reduces to the unfolded version. This allows to
-precisely mimic the usual unfolding rules of :term:`constants <constant>`. Projections
-obey the usual ``simpl`` flags of the ``Arguments`` command in particular.
+However, to take the δ flag into account, projections can be in two states:
+folded or unfolded. An unfolded primitive projection application obeys the rule
+above, while the folded version delta-reduces to the unfolded version. This
+allows to precisely mimic the usual unfolding rules of :term:`constants <constant>`. Projections obey the usual ``simpl`` flags of the :cmd:`Arguments`
+command in particular.
```
```suggestion
However, to take the δ flag into account, projections can be in two states:
folded or unfolded. An unfolded primitive projection application obeys the rule
above, while the folded version delta-reduces to the unfolded version. This
allows to precisely mimic the usual unfolding rules of :term:`constants <constant>`.
Projections obey the usual ``simpl`` flags of the :cmd:`Arguments`
command in particular.
```

long line.  (The wording is not great, but let's not touch that today.)

jfehrle commented on doc/sphinx/language/core/records.rst:
```diff
@@ -305,17 +311,16 @@ from folded ones.
-Additionally, user-written :g:`match` constructs on primitive records
-are desugared into substitution of the projections, they cannot be
-printed back as :g:`match` constructs.
+To ease compatibility with ordinary record types, each primitive projection is
+also defined as an ordinary :term:`constant` taking parameters and an object of
+the record type as arguments, and whose :term:`body` is an application of the
+unfolded primitive projection of the same name. These constants are used when
+elaborating partial applications of the projection. One can distinguish them
+from applications of the primitive projection if the :flag:`Printing Primitive
+Projection Parameters` flag is off: For a primitive projection application,
+parameters are printed as underscores while for the compatibility projections
+they are printed as usual.
+
+Additionally, user-written :g:`match` constructs on primitive records are
+desugared into substitution of the projections, they cannot be printed back as
+:g:`match` constructs.
```
This seems to be only a reformatting without a user-visible wording change.  While it's reasonable to split really long lines, reformatting in this case makes for more work in the review (i.e., figuring out whether there is a wording change) and makes it harder to identify the commits that made substantive changes.  Better not to change this now.

Alizter replied:
I have not changed the wording here, only the formatting. I should have perhaps done the formatting in a later commit and squashed after review.

Alizter replied:
The reason I have done these, is because I was splitting up a large line. In my IDE (VSCode) I just press a shortcut and it hard wraps the entire paragraph. I'll try to keep the diff clearer next time.

jfehrle replied:
> I have not changed the wording here, only the formatting. I should have perhaps done the formatting in a later commit and squashed after review.

Slightly better, but I would not reformat even in a separate commit absent a good reason to do so.  (And the principle applies even more strongly to source code.)

jfehrle reviewed (commented):
A few more suggestions

Alizter commented on doc/sphinx/language/core/records.rst:
```diff
@@ -53,8 +53,8 @@ expressions. In this sense, the :cmd:`Record` construction allows defining
-   :attr:`universes(template)`, :attr:`universes(cumulative)`, and
-   :attr:`private(matching)` attributes.
+   :attr:`universes(template)`, :attr:`universes(cumulative)`,
+   :attr:`private(matching)`, and :attr:`projections(primitive)` attributes.
```
Huh, I had no idea! :-)

cpitclaudel replied:
You got this backwards, @jfehrle: the Oxford comma is typical in AmE and somewhat less common in BrE; see https://en.wikipedia.org/wiki/Serial_comma

Alizter replied:
OK now I am confused. I actually agree with @jfehrle on the comma convention here, since it is what I am used too. However, before I changed anything in this document, the attributes were listed with the extra `,`. Which one should be ought to use?

jfehrle replied:
@cpitclaudel Wikipedia is mistaken.  The Oxford comma is not commonly used in US English--not in journalism, not in general-interest books and not in the few academic papers I've read.  Nor have I have seen it much in the Coq documentation.

@Zimmi48 Any thoughts?

Zimmi48 replied:
> Wikipedia is mistaken.

I didn't check the references, but I'm not expecting this claim to be wrong:

>A majority of American style guides mandate use of the serial comma. These include: APA style,[8] The Chicago Manual of Style, Garner's Modern American Usage,[9] The MLA Style Manual, Strunk and White's The Elements of Style,[10] and the U.S. Government Printing Office Style Manual. By contrast, the Associated Press Stylebook and The New York Times Style Book[11] advise against it.

Also, for the record, by own grammar checking tool, which is set to American English also recommends it.

That being said, since this is apparently a debated topic, I would tend to side with Jim against the Oxford comma. But I would also not lose much time over this.

cpitclaudel replied:
@jfehrle Well, I'm no expert, so I usually go by published sources; here's what my copy of "Garner's Modern English Usage" has to say (sorry, I don't have an electronic copy):

![image](https://user-images.githubusercontent.com/8181630/127564792-f5ad7da3-be63-4beb-8ea1-29660d91e301.png)

![image](https://user-images.githubusercontent.com/8181630/127564963-c013700f-779d-413b-8538-cc7d0a3d85f9.png)

FWIW, Coq is copyrighted to `The Coq development team, INRIA, CNRS, and contributors`

jfehrle replied:
@cpitclaudel I'm familiar with the discussion.  It's just that the Oxford comma is rarely used in practice.  As Wikipedia notes, the US style guides are not unanimous.  Actual usage trumps any "authoritative" source.

The part about "When the members are compound, calling for _and_ within themselves, clarity demands the final comma." would be the main exception in my experience.

cpitclaudel replied:
Eh, we have actual usage right in front of us: the way the manual is currently written.  So I see no need to change it to conform to perceptions about what's more or less commonly used out there.

Alizter replied:
I had a look back through the git blame and it appears that @Zimmi48 wrote it first in the style @jfehrle suggests, it seems later @ejgallego changed it to have the `,` before the `and`.

jfehrle replied:
@cpitclaudel I would not select the documentation as an exemplar of excellent, well-edited writing from which you can infer what is common usage.

cpitclaudel replied:
Indeed not, and hence there's no need to get into these long arguments about comma usage preferences :) Sorry that I got dragged into this — I was originally reacting only to the comment that this would "generally be considered incorrect", which isn't true, AFAICT.

jfehrle replied:
Indeed, "incorrect" was a bit too emphatic.

Zimmi48 replied:
I suggest settling for inconsistency on this topic, that is, let the PR author do what they want, and avoid large-scale changes relative to these commas.

Zimmi48 replied:
> I had a look back through the git blame and it appears that @Zimmi48 wrote it first in the style @jfehrle suggests, it seems later @ejgallego changed it to have the `,` before the `and`.

That is: there is no Oxford comma in all copyright statements (manual and header files) but they appear in the author fields of the `.opam` files. Not really a "change" IMHO.

Alizter commented on doc/sphinx/language/core/records.rst:
```diff
@@ -79,7 +79,7 @@ in which case the correctness of :n:`@type__3` may rely on the instance :n:`@ter
    and ``Rat_irred_cond`` depends on both ``top`` and ``bottom``.
 
-Let us now see the work done by the ``Record`` macro. First the macro
+Let us now see the work done by the :cmd:`Record` macro. First the macro
```
I'll check for them.

Alizter replied:
Alright, replaced two more occurances.

Alizter commented on doc/sphinx/language/core/records.rst:
```diff
@@ -235,68 +236,73 @@ the errors of inductive definitions, as described in Section
+   These kind of record types differ from their traditional versions in the
+   sense that dependent
    elimination is not available for them and only non-dependent case analysis
    can be defined.
```
This seems to be the case:
```coq

#[projections(primitive)]
Inductive foo := {
 bar : Type;
}.
```
yields
```
foo : Type

foo is not universe polymorphic
foo has primitive projections without eta conversion.
Expands to: Inductive Top.foo
```
I believe the `Inductive` command disallows eta conversion to take place.

Alizter commented:
@coq/vernac-maintainers Anyone willing to assign?

maximedenes reviewed (approved):
Looks good, thanks! At some point, we may consider packing all these configuration parameters  in a record to keep function signatures readable.

Alizter commented:
Yep, a similar idea crossed my mind but I decided it would be too disruptive for this PR. I'll create an issue for it.

maximedenes commented:
@coqbot: merge now

maximedenes commented:
@coqbot: merge now
